appsec/config.go

```diff
@@ -0,0 +1,174 @@
+
+// Configuration constants and default values
+const (
+	defaultAPISecSampleRate     = 10. / 100
+	defaultObfuscatorKeyRegex   = `(?i)(?:p(?:ass)?w(?:or)?d|pass(?:_?phrase)?|secret|(?:api_?|private_?|public_?)key)|token|consumer_?(?:id|key|secret)|sign(?:ed|ature)|bearer|authorization`
+	defaultObfuscatorValueRegex = `(?i)(?:p(?:ass)?w(?:or)?d|pass(?:_?phrase)?|secret|(?:api_?|private_?|public_?|access_?|secret_?)key(?:_?id)?|token|consumer_?(?:id|key|secret)|sign(?:ed|ature)?|auth(?:entication|orization)?)(?:\s*=[^;]|"\s*:\s*"[^"]+")|bearer\s+[a-z0-9\._\-]+|token:[a-z0-9]{13}|gh[opsu]_[0-9a-zA-Z]{36}|ey[I-L][\w=-]+\.ey[I-L][\w=-]+(?:\.[\w.+\/=-]+)?|[\-]{5}BEGIN[a-z\s]+PRIVATE\sKEY[\-]{5}[^\-]+[\-]{5}END[a-z\s]+PRIVATE\sKEY|ssh-rsa\s*[a-z0-9\/\.+]{100,}`
+	defaultWAFTimeout           = 4 * time.Millisecond
+	defaultTraceRate            = uint(100) // up to 100 appsec traces/s
+)
+
+// APISecConfig holds the configuration for API Security schemas reporting
+// It is used to enabled/disable the feature as well as to configure the rate
+// at which schemas get reported,
+type APISecConfig struct {
+	Enabled    bool
+	SampleRate float64
+}
+
+// ObfuscatorConfig wraps the key and value regexp to be passed to the WAF to perform obfuscation.
+type ObfuscatorConfig struct {
+	KeyRegex   string
+	ValueRegex string
+}
+
+// NewAPISecConfig creates and returns a new API Security configuration by reading the env
+func NewAPISecConfig() APISecConfig {
+	return APISecConfig{
+		Enabled:    apiSecurityEnabled(),
+		SampleRate: readAPISecuritySampleRate(),
+	}
+}
+
+func apiSecurityEnabled() bool {
+	enabled, _ := strconv.ParseBool(os.Getenv(envAPISecEnabled))
+	return enabled
+}
+
+func readAPISecuritySampleRate() float64 {
+	value := os.Getenv(envAPISecSampleRate)
+	rate, err := strconv.ParseFloat(value, 64)
+	if err != nil {
+		logEnvVarParsingError(envAPISecSampleRate, value, err, defaultAPISecSampleRate)
+		return defaultAPISecSampleRate
+	}
+	if rate < 0. || rate > 1. {
+		logUnexpectedEnvVarValue(envAPISecSampleRate, rate, "value must be between 0 and 1", defaultAPISecSampleRate)
+		return defaultAPISecSampleRate
+	}
+	return rate
+}
+
+// NewObfuscatorConfig creates and returns a new WAF obfuscator configuration by reading the env
+func NewObfuscatorConfig() ObfuscatorConfig {
+	keyRE := readObfuscatorConfigRegexp(envObfuscatorKey, defaultObfuscatorKeyRegex)
+	valueRE := readObfuscatorConfigRegexp(envObfuscatorValue, defaultObfuscatorValueRegex)
+	return ObfuscatorConfig{KeyRegex: keyRE, ValueRegex: valueRE}
+}
+
+func readObfuscatorConfigRegexp(name, defaultValue string) string {
+	val, present := os.LookupEnv(name)
+	if !present {
+		log.Debug("appsec: %s not defined, starting with the default obfuscator regular expression", name)
+		return defaultValue
+	}
+	if _, err := regexp.Compile(val); err != nil {
+		logUnexpectedEnvVarValue(name, val, "could not compile the configured obfuscator regular expression", defaultValue)
+		return defaultValue
+	}
```

Comment:
Are we certain the regex dialect used by the WAF and the one used by Go's standard library are identical? If not, this check might allow regexes that the WAF won't understand, or might reject regexes the WAF would have accepted...

Reply:
We are, both use the RE2 syntax :
- [Go](https://pkg.go.dev/regexp)
- [WAF](https://github.com/DataDog/libddwaf/blob/f1dc07f8f4cb328cc080aa927f6b0d1e0e9a7aa9/src/obfuscator.hpp#L10)